### DISCUSSION TITLE
Passing in runtime data on model instantiation

### DISCUSSION THREAD

**Discussion by aboyer1013** [xaviergonz/mobx-keystone#392] [Q&A] [answered]
How do I pass runtime data when instantiating models?

1.  Currently, I have to define my own setter and then set it afterwards.
2.  This also means I have to always do non-null assertions to make TypeScript happy.

Is there a better way to do this?

[Code Sandbox example](https://codesandbox.io/s/jolly-burnell-1h8rg?file=/src/index.ts)

```
@model('MyModel')
class MyModel extends Model({
  id: idProp,
}) {
  fooAction!: () => void

  @modelAction
  setFooAction(fooAction: () => void) {
    this.fooAction = fooAction;
  }
});

const myModel = new MyModel({
  // Wanting to pass runtime data here, but it throws an error
  // fooAction: () => { console.log('fooAction init') },
});

myModel.setFooAction(() => { console.log('fooAction')});
myModel.fooAction();
```

**Comment by sisp** [comment, marked as answer]
What you're trying to do might not be sound. If you set `fooAction` upon model instantiation, then created a snapshot, and later restored the state from the snapshot again, `fooAction` would not be set then because it's neither part of the snapshot (because (a) it's not a model prop and (b) functions aren't snapshotable/serializable anyway) nor part of the model definition.

[Contexts](https://mobx-keystone.js.org/contexts) might be what you're looking for. You can inject any kind of data into a state tree via a context and retrieve it at a child node, e.g. in your model instance.

Just FYI, you don't need to decorate `setFooAction` with `@modelAction` because it doesn't assign a value to a model prop.

**Reply by aboyer1013** [reply]
Thanks for the feedback!

I glossed over the last bit in the context documentation where it explains how to apply volatile (runtime) data at construction time. That should suit my needs just fine.
